dali/pipeline/operator/builtin/external_source.cu

```diff
@@ -27,41 +27,47 @@ struct ExternalSource<GPUBackend>::RecycleFunctor {
     assert(!"Should never happen");
   }
   RecycleFunctor(RecycleFunctor &&) = default;
+  RecycleFunctor& operator=(const RecycleFunctor&) = default;
+  RecycleFunctor& operator=(RecycleFunctor&&) = default;
+  ~RecycleFunctor() = default;
 
-  RecycleFunctor(
-      ExternalSource<GPUBackend> *owner,
-      std::list<uptr_cuda_event_type> event,
-      std::list<uptr_tl_type> ptr)
-  : owner(owner), event(std::move(event)), ptr(std::move(ptr)) {}
+
+  RecycleFunctor(ExternalSource<GPUBackend> *owner, std::list<uptr_cuda_event_type> event,
+                 std::list<uptr_tl_type> ptr, std::list<uptr_cuda_event_type> internal_copy_to_gpu)
+          : owner(owner), event(std::move(event)), copy_to_gpu(std::move(internal_copy_to_gpu)),
+            ptr(std::move(ptr)) {}
 
   ExternalSource<GPUBackend> *owner;
-  std::list<uptr_cuda_event_type> event;
+  std::list<uptr_cuda_event_type> event, copy_to_gpu;
   std::list<uptr_tl_type> ptr;
   void operator()() {
-    owner->RecycleBuffer(ptr, &event);
+    owner->RecycleBuffer(ptr, &event, &copy_to_gpu);
   }
 };
 
 template<>
 void ExternalSource<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
   std::list<uptr_tl_type> data;
-  std::list<uptr_cuda_event_type> cuda_event;
+  std::list<uptr_cuda_event_type> cuda_event, internal_copy_to_gpu;
   {
     std::unique_lock<std::mutex> busy_lock(busy_m_);
     cv_.wait(busy_lock, [&data = data_in_tl_]{return !data.empty();});
     auto is_data_in_tl = data_in_tl_.front();
     data_in_tl_.pop_front();
     DALI_ENFORCE(is_data_in_tl, "Cannot feed non-contiguous data to GPU op.");
     data = tl_data_.PopFront();
+    internal_copy_to_gpu = copy_to_gpu_events_.PopFront();
     cuda_event = cuda_events_.GetEmpty();
   }
 
+  cudaEventSynchronize(*internal_copy_to_gpu.front());
```

Comment:
Why not something like `cudaStreamWaitEvent`? Do we need this busy wait here?

Reply:
Done